src/pages/studyView/charts/barChart/BarChartToolTip.tsx

```diff
@@ -0,0 +1,110 @@
+import * as React from "react";
+import * as _ from 'lodash';
+import {computed} from "mobx";
+import { IAlterationData, IAlterationCountMap } from "pages/resultsView/cancerSummary/CancerSummaryContent";
+import { Popover } from "react-bootstrap";
+import * as ReactDOM from "react-dom";
+import classnames from "classnames";
+
+export type ToolTipRange = {
+    start: number | undefined,
+    end: number | undefined,
+    special: string | undefined
+};
+
+export type BarChartToolTipProps = {
+    mousePosition: { x: number, y: number },
+    totalBars: number,
+    currentBarIndex: number,
+    range: ToolTipRange | null,
+    sampleCount: number,
+    windowWidth: number,
+};
+
+const OFFSET = 17;
+const WIDTH = 150;
+
+export default class BarChartToolTip extends React.Component<BarChartToolTipProps, {}> {
+
+    /**
+     * When the active bar is past the middle of the plot, render on the left side of the bar
+     */
+    @computed
+    get placement(): "left" | "right" {
+        return (this.props.totalBars < this.props.currentBarIndex * 2) ? "left" : "right";
+    }
+
+    @computed
+    get positionLeft(): number {
+        return this.props.mousePosition.x + (this.placement == "left" ? (-OFFSET - WIDTH) : OFFSET); 
+    }
+
+    @computed
+    get transform(): string | undefined {
+        return (this.placement === "left" ? "translate(-100%,0%)" : undefined)
+    }
+
+    translateSpecialText(text: string | undefined): string {
+        if (!text) {
+            return "";
+        }
+
+        if (text === ">=") {
+            return "≥";
+        }
+
+        if (text === "<=") {
+            return "≤";
+        }
+
+        return text;
+    }
+
+
+    @computed
+    get rangeText(): string {
+        if (!this.props.range) {
+            return "";
+        }
+
+        const special = this.translateSpecialText(this.props.range.special);
+        
+        if (this.props.range.start === undefined) {
+            if (this.props.range.end === undefined) {
+                return special;
+            } else {
+                return `${special}${this.props.range.end}`
+            }
+        } else {
+            if (this.props.range.end === undefined) {
+                return `${special}${this.props.range.start}`
+            } else {
+                return `${this.props.range.start} - ${this.props.range.end}`
+            }
+        }
+    }
+
+    
+    render() {
+        if (!this.props.range) {
+            return null;
+        }
+
+        return <Popover
+            arrowOffsetTop={OFFSET}
```

Comment:
I'm not sure what the consideration is to come up with the offset. We also some offset settings for the ScatterPlot in the component `ScatterPlotTooltip`. May worth a discussion to consolidate them.

Reply:
Hmm. I took this value from `CancerSummaryChart`. I looked at `ScatterPlotTooltip` and I don't think those values would be appropriate for a bar chart - each bar is only `20px` wide, so a `25px` would point to the wrong bar.

That said, you are right that I borrowed a value and didn't consolidate. I will convert the offsets in `CancerSummaryChart` to constants and reuse them.